src/main/java/com/lgtm/easymoney/services/impl/LoanServiceImpl.java

```diff
@@ -0,0 +1,190 @@
+    String desc = req.getDescription();
+    User borrower = userService.getUserById(req.getFromUid());
+    User lender = userService.getUserById(req.getToUid());
+    // validate if loan user type is valid
+    validateLoanUsers(borrower, lender);
+    // create loan request
+    Transaction loan = createLoanRequest(borrower, lender, amount, desc,
+        Category.valueOf(category.toUpperCase()));
+    // response
+    return loan == null ? new ResourceCreatedRsp(null) : new ResourceCreatedRsp(loan.getId());
+  }
+
+  @Override
+  public LoanRsp getLoansByUid(Long uid) {
+    User user = userService.getUserById(uid);
+    List<Transaction> loans = getLoansByUser(user);
+    // response
+    LoanRsp response = new LoanRsp();
+    response.setSuccess(true);
+    List<TransactionRsp> transferRsps =
+        transactionService.generateListResponseFromTransactions(loans);
+    response.setLoans(transferRsps);
+    response.setMessage("User's loans returned");
+    return response;
+  }
+
+  private boolean approveLoan(Transaction loan) {
+    loan.setStatus(TransactionStatus.TRANS_PENDING);
+    transactionService.saveTransaction(loan);
+    boolean success = transactionService.executeTransaction(loan);
+    if (success) {
+      loan.setStatus(TransactionStatus.LOAN_APPROVED);
+      transactionService.saveTransaction(loan);
+    } else {
+      loan.setStatus(TransactionStatus.LOAN_PENDING);
+      transactionService.saveTransaction(loan);
+    }
+    return success;
+  }
+
+  @Override
+  public LoanRsp approveLoan(RequestAcceptDeclineReq req) {
+    // get params
+    Long fromUid = req.getFromUid();
+    Long toUid = req.getToUid();
+    Long lid = req.getRequestid();
+    // validate
+    User lender = userService.getUserById(fromUid);
+    User borrower = userService.getUserById(toUid);
+    validateLoanUsers(borrower, lender);
+    if (!validateLoanRequest(lid, fromUid, toUid)) {
+      throw new InvalidUpdateException("approve loan", lid, "loanId", lid);
+    }
+    // approve loan and transfer money
+    Transaction loan = getLoanById(lid);
+    boolean success = approveLoan(loan);
+    if (!success) {
+      throw new InvalidUpdateException("approve loan", loan.getId(),
+          "loan id", loan.getId());
+    }
+    // create a loan-payback request from lender to borrower, storing loanId as description
+    Transaction payback = requestService.createRequest(lender, borrower, loan.getAmount(),
+        String.valueOf(loan.getId()), Category.LOAN_PAYBACK);
+    // response returns the approved loan and the loan-payback request
+    LoanRsp response = new LoanRsp();
+    response.setSuccess(true);
+    List<TransactionRsp> transferRsps =
+        transactionService.generateListResponseFromTransactions(List.of(loan, payback));
+    response.setLoans(transferRsps);
+    response.setMessage("User's approved loan and required payback request returned");
+    return response;
+  }
+
+  private boolean declineLoan(Transaction loan) {
```

Comment:
i can see it's modified from transactions, but i was thinking if it's better to return more meaningful data instead of success flag? reason is exceptions are well-handled already.

Reply:
So probably just return the loan json data?

Reply:
sounds good. or should we modify the transaction service as well to return `repo.save(t)`? @jenniferduan45

Reply:
Changed it in loanServiceImpl and will push later. But if we change requestServiceImpl now, we also need to change its unit test.Probably change it in another PR?